OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt

```diff
@@ -167,6 +193,72 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
         })
     }
 
+    private fun initializedSyncContainer() {
+        binding.switchSync.setOnCheckedChangeListener { buttonView, isChecked ->
+            environment.analyticsRegistry.trackCourseDatesCalendarToggleButton(courseData.courseId, courseData.mode, isSelfPaced, isChecked)
+            if (isChecked) {
+                if (!permissions.any { permission -> PermissionsUtil.checkPermissions(permission, contextOrThrow) }) {
+                    val title: String = ResourceUtil.getFormattedString(resources, R.string.title_request_calendar_permission, AppConstants.PLATFORM_NAME, environment.config.platformName).toString()
+                    val message: String = ResourceUtil.getFormattedString(resources, R.string.message_request_calendar_permission, AppConstants.PLATFORM_NAME, environment.config.platformName).toString()
+
+                    AlertDialogFragment.newInstance(title, message, getString(R.string.label_ok),
+                            { dialogInterface: DialogInterface, i: Int ->
+                                PermissionsUtil.requestPermissions(PermissionsUtil.CALENDAR_PERMISSION_REQUEST, permissions, this@CourseDatesPageFragment)
+                            },
+                            getString(R.string.label_do_not_allow),
+                            { dialogInterface: DialogInterface?, i: Int ->
+                                environment.analyticsRegistry.trackCourseDatesCalendarPermission(courseData.courseId, courseData.mode, isSelfPaced, false)
+                                buttonView.isChecked = false
+                            }).show(childFragmentManager, null)
+
+                } else {
+                    askForCalendarSync()
+                }
+            } else {
+                val calendarId = getCalenderId(accountName)
+                if (calendarId != (-1).toLong()) {
+                    deleteCalendar(calendarId)
+                }
+            }
+        }
+    }
+
+    private fun askForCalendarSync() {
+        val title: String = ResourceUtil.getFormattedString(resources, R.string.title_add_course_calendar, AppConstants.COURSE_NAME, courseName).toString()
+        val message: String = ResourceUtil.getFormattedString(resources, R.string.message_add_course_calendar, AppConstants.COURSE_NAME, courseName).toString()
+
+        AlertDialogFragment.newInstance(title, message, getString(R.string.label_ok),
+                { dialogInterface: DialogInterface, i: Int ->
+                    environment.analyticsRegistry.trackCourseDatesCalendarAddDates(courseData.courseId, courseData.mode, isSelfPaced, true)
+                    insertCalendarEvent()
```

Comment:
It should be async, cuz this moves the app in a wait busy state for lower-end devices like Nexus-5.